native/explorer/src/datatypes.rs

```diff
@@ -352,12 +359,105 @@ impl From<NaiveDateTime> for ExDateTime {
     }
 }
 
-impl Literal for ExDateTime {
+impl Literal for ExNaiveDateTime {
     fn lit(self) -> Expr {
         NaiveDateTime::from(self).lit()
     }
 }
 
+#[derive(NifStruct, Copy, Clone, Debug)]
+#[module = "DateTime"]
+pub struct ExDateTime<'a> {
+    pub calendar: Atom,
+    pub day: u32,
+    pub hour: u32,
+    pub microsecond: (u32, u32),
+    pub minute: u32,
+    pub month: u32,
+    pub second: u32,
+    pub std_offset: i64,
+    pub time_zone: &'a str,
+    pub utc_offset: i64,
+    pub year: i32,
+    pub zone_abbr: &'a str,
+}
+
+// impl From<i64> for ExDateTime<'_> {
+//     fn from(microseconds: i64) -> Self {
+//         timestamp_to_datetime_utc(microseconds).into()
+//     }
+// }
+
+impl From<ExDateTime<'_>> for i64 {
+    fn from(dt: ExDateTime<'_>) -> i64 {
+        let duration = NaiveDate::from_ymd_opt(dt.year, dt.month, dt.day)
+            .unwrap()
+            .and_hms_micro_opt(dt.hour, dt.minute, dt.second, dt.microsecond.0)
+            .unwrap()
+            .signed_duration_since(
+                NaiveDate::from_ymd_opt(1970, 1, 1)
+                    .unwrap()
+                    .and_hms_opt(0, 0, 0)
+                    .unwrap(),
+            );
+
+        match duration.num_microseconds() {
+            Some(us) => us,
+            None => duration.num_milliseconds() * 1_000,
+        }
+    }
+}
+
+// TODO-BILLY: finish this.
+impl<'a> From<DateTime<Tz>> for ExDateTime<'a> {
+    fn from(dt_tz: DateTime<Tz>) -> ExDateTime<'a> {
+        let & time_zone = dt_tz.offset().tz_id();
+        let & zone_abbr = dt_tz.offset().abbreviation();
```

Comment:
@philss (or anyone!) I've hit the limit of my ability to cargo cult Rust code 😭.

The `ExDateTime` struct currently looks like this:

```rust
#[derive(NifStruct, Copy, Clone, Debug)]
#[module = "DateTime"]
pub struct ExDateTime<'a> {
    // ...
    pub time_zone: &'a str,
    // ...
    pub zone_abbr: &'a str,
}
```

I need to derive these fields from the `DateTime` object, so they're not known at compile time. AFAICT I've got two options: they can have type `&'a str` or `String`. I can implement this if they have type `String`, but then we can no longer `#[derive(Copy)]`. So I've tried to implement it with `&'a str`, but I haven't figured out how yet.

Any pointers would be appreciated!

Reply:
Oh man, I really don't know how to solve this easily. Sorry :/
Do we need to support `Copy`? I would try to use `String` if we don't need this.

Reply:
[8f5be24](https://github.com/elixir-explorer/explorer/pull/903/commits/8f5be2464294275774e187fe991b60e6d3b6ed18)

Ok, so we _did_ end up needing `Copy` because of how I wrote `s_from_list_datetime`. Specifically, this breaks:

```rust
val: Vec<Option<ExDateTime>>
// ...
val.iter()
    .map(|dt| dt.map(|dt| dt.into()))
//            ^^^^^^^^^^^^^^^^^^^^^^ this implicitly requires Copy
    .collect::<Vec<Option<i64>>>(),
```

I've reverted to `&'a str` for now since all the tests pass. I suspect something we try with `%DateTime{}` will eventually break because of this, but I say let's address it then.

Thoughts?

Reply:
No thoughts for now :/
Let's go with this version, and we improve in the future :D